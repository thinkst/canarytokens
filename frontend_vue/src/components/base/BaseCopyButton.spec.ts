import { mount } from '@vue/test-utils';
import { vi } from 'vitest';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faCopy, faCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import BaseCopyButton from './BaseCopyButton.vue';

library.add(faCopy, faCheck);
const tooltip = vi.fn();

let copied = false;

const mockCopy = vi.fn(() => {
  copied = true;
});

// Not sure this is the best way to mock the useClipboard hook
// mockCopy seems to not update copied correctly
vi.mock('@vueuse/core', () => ({
  useClipboard: vi.fn().mockImplementation(() => ({
    isSupported: true,
    copy: mockCopy,
    copied: copied,
  })),
}));

describe('BaseCopyButton', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it('is called', () => {
    const wrapper = mount(BaseCopyButton, {
      props: {
        content: 'Test content',
      },
      global: {
        stubs: { FontAwesomeIcon },
      },
      directives: {
        tooltip,
      },
    });
    expect(wrapper.exists()).toBeTruthy();
  });

  it('render correctly', () => {
    const wrapper = mount(BaseCopyButton, {
      props: {
        content: 'Test content',
      },
      global: {
        stubs: { FontAwesomeIcon },
      },
      directives: {
        tooltip,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('triggers Copy on click', async () => {
    const onClick = vi.fn();

    const wrapper = mount(BaseCopyButton, {
      props: {
        content: 'Test content',
      },
      attrs: { onClick: onClick },
      global: {
        stubs: { FontAwesomeIcon },
      },
      directives: {
        tooltip,
      },
    });
    await wrapper.find('button').trigger('click');
    expect(mockCopy).toHaveBeenCalledTimes(1);
  });

  it('displays the correct initial tooltip text', () => {
    const wrapper = mount(BaseCopyButton, {
      props: {
        content: 'Test content',
      },
      global: {
        stubs: { FontAwesomeIcon },
      },
      directives: {
        tooltip,
      },
    });
    //@ts-ignore
    const tooltipText = wrapper.vm.tooltipText;
    expect(tooltipText).toBe('Copy to clipboard');
  });

  it('displays the correct tooltip text after click', async () => {
    const wrapper = mount(BaseCopyButton, {
      props: {
        content: 'Test content',
      },
      global: {
        stubs: { FontAwesomeIcon },
      },
      directives: {
        tooltip,
      },
    });
    await wrapper.find('button').trigger('click');
    await vi.advanceTimersByTime(300);
    //@ts-ignore
    const tooltipText = wrapper.vm.tooltipText;
    expect(tooltipText).toBe('Copied!');
  });
});
