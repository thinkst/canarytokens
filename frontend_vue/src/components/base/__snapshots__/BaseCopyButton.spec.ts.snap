// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseCopyButton > render correctly 1`] = `
"<button data-v-e7e863f3="" class="h-[2rem] w-[2rem] font-semibold text-white rounded-full bg-green hover:bg-green-300 transition duration-100" aria-label="Copy to clipboard">
  <transition-stub data-v-e7e863f3="" name="fade" mode="out-in" appear="false" persisted="false" css="true"><svg data-v-e7e863f3="" class="svg-inline--fa fa-copy" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="copy" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
      <path class="" fill="currentColor" d="M208 0H332.1c12.7 0 24.9 5.1 33.9 14.1l67.9 67.9c9 9 14.1 21.2 14.1 33.9V336c0 26.5-21.5 48-48 48H208c-26.5 0-48-21.5-48-48V48c0-26.5 21.5-48 48-48zM48 128h80v64H64V448H256V416h64v48c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V176c0-26.5 21.5-48 48-48z"></path>
    </svg></transition-stub>
  <!--v-if-->
</button>"
`;
